### PR TITLE
Compute required beatmap attributes in DifficultyCalculators

### PR DESCRIPTION
**Pull request by smoogipoo**
The primary goal of this set of changes is to output beatmap attributes from the difficulty calculators for the performance calculator to consume. The reason for this is two ways:
1. The diffcalc server will be able to overwrite all attributes currently in the DB.
2. An osu!lazer-based performance calculator at some point in the future will not need to parse the beatmap, and instead be able to read attributes from the DB as is currently done in osu-performance.

### CODE REVIEW

**Comment by Tom94** [comment]
I would like to comment on your second point: I think it'd actually be nice to have the performance calculator calculate difficulties on-the-fly as they are required (with caching). This would have the nice benefit of allowing us to use more complex difficulty metrics for performance computation than can be stored in the DB, such as per-object strain. I have a few idea on how such data could be useful without requiring scores to have per-hitobject data.

**Comment by smoogipoo** [comment]
Might be useful indeed. I'm fine with removing my change if that's the path we should go down (perf calc always computing dificulty).

File: osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs [hunk 1]
```diff
@@ -32,10 +32,6 @@ protected virtual void ApplyMods(Mod[] mods)
             TimeRate = clock.Rate;
         }
 
-        protected virtual void PreprocessHitObjects()
-        {
-        }
-
-        public abstract double Calculate(Dictionary<string, double> categoryDifficulty = null);
+        public abstract double Calculate(Dictionary<string, object> categoryDifficulty = null);
```
Comment (peppy) [inline osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs]: This could do with some documentation (what is the nullable dictionary parameter?)

**Comment by smoogipoo** [comment]
Closing for re-implementation of difficulty calculators.

**Pull request closed without merging**